File: vc-agent/docs/diagrams/rough.py
"""Hand-drawn SVG primitives, in the Excalidraw manner.

### Why this exists rather than an SVG filter

The obvious way to get a sketchy look is `feTurbulence` + `feDisplacementMap`
over the shapes. It renders correctly in a browser window and is **silently
dropped when Chrome prints to PDF** — the boxes come out perfectly straight and
nothing warns you. Measured on Chrome 148 against this document: filtered and
unfiltered output were pixel-identical in the PDF.

So the wobble is baked into the path data instead. A rough rectangle here is a
real `<path>` with jittered corners and slightly bowed edges; nothing at render
time has to cooperate.

### What makes it read as hand-drawn

Three things, and the third is the one people miss:

1. **Corners miss.** Each vertex moves by a pixel or two.
2. **Edges bow.** A drawn line is never straight; each edge is a quadratic curve
   with a small perpendicular bulge.
3. **Every stroke is drawn twice**, with different jitter. This is Excalidraw's
   signature and the single biggest contributor — a pen goes back over a line.

### Determinism

The jitter comes from a seeded PRNG keyed per shape, so regenerating the document
produces byte-identical output. A diagram that reshuffles itself on every build
makes every rebuild look like a change in review.
"""

from __future__ import annotations

import math
import random

# Excalidraw's palette, near enough. Fills are the pale end of each hue.
INK = "#1e1e1e"
PALETTE = {
    "blue":   ("#1971c2", "#e7f5ff"),
    "green":  ("#2f9e44", "#ebfbee"),
    "red":    ("#c92a2a", "#fff5f5"),
    "orange": ("#e8590c", "#fff4e6"),
    "violet": ("#5f3dc4", "#f8f0fc"),
    "grey":   ("#868e96", "#f8f9fa"),
    "ink":    (INK, "#ffffff"),
}
HAND = "Comic Sans MS, Comic Neue, DejaVu Sans, sans-serif"


class Pen:
    """A seeded source of wobble. One Pen per figure keeps figures independent."""

    def __init__(self, seed: int, roughness: float = 1.0) -> None:
        self._rng = random.Random(seed)
        self.roughness = roughness

    def j(self, amount: float = 1.6) -> float:
        return self._rng.uniform(-amount, amount) * self.roughness

    # ---------------------------------------------------------------- edges

    def _edge(self, x1: float, y1: float, x2: float, y2: float, bow: float) -> str:
        """One bowed edge as a quadratic curve, endpoints jittered."""
        dx, dy = x2 - x1, y2 - y1
        length = math.hypot(dx, dy) or 1.0
        # Perpendicular unit vector — the bulge goes across the line, not along it.
        nx, ny = -dy / length, dx / length
        swell = self.j(bow) * min(1.0, length / 90)
        mx = (x1 + x2) / 2 + nx * swell
        my = (y1 + y2) / 2 + ny * swell
        return (
            f"M{x1 + self.j():.1f},{y1 + self.j():.1f} "
            f"Q{mx:.1f},{my:.1f} {x2 + self.j():.1f},{y2 + self.j():.1f}"
        )

    # ---------------------------------------------------------------- shapes

    def rect(self, x, y, w, h, colour="ink", *, width=1.6, fill=True, dash="") -> str:
        stroke, wash = PALETTE[colour]
        out = []
        if fill:
            # The fill is a plain polygon set slightly inside the outline, so the
            # wobbly stroke reads as drawn *over* a wash rather than clipping it.
            out.append(
                f'<path d="M{x+1},{y+1} L{x+w-1},{y+1} L{x+w-1},{y+h-1} L{x+1},{y+h-1} Z" '
                f'fill="{wash}" stroke="none"/>'
            )
        corners = [(x, y), (x + w, y), (x + w, y + h), (x, y + h)]
        extra = f' stroke-dasharray="{dash}"' if dash else ""
        for _pass in range(2):
            d = " ".join(
                self._edge(*corners[i], *corners[(i + 1) % 4], bow=1.8)
                for i in range(4)
            )
            out.append(
                f'<path d="{d}" fill="none" stroke="{stroke}" '
                f'stroke-width="{width}" stroke-linecap="round"{extra}/>'
            )
        return "".join(out)

    def circle(self, cx, cy, r, colour="ink", *, width=1.5, fill=True, dash="") -> str:
        stroke, wash = PALETTE[colour]
        out = []
        if fill:
            out.append(f'<circle cx="{cx}" cy="{cy}" r="{r - 0.6}" fill="{wash}" stroke="none"/>')
        extra = f' stroke-dasharray="{dash}"' if dash else ""
        for _pass in range(2):
            pts = []
            steps = 11
            for i in range(steps + 1):
                a = 2 * math.pi * i / steps
                rr = r + self.j(1.1)
                pts.append((cx + rr * math.cos(a), cy + rr * math.sin(a)))
            d = f"M{pts[0][0]:.1f},{pts[0][1]:.1f} " + " ".join(
                f"L{px:.1f},{py:.1f}" for px, py in pts[1:]
            ) + " Z"
            out.append(
                f'<path d="{d}" fill="none" stroke="{stroke}" '
                f'stroke-width="{width}" stroke-linejoin="round"{extra}/>'
            )
        return "".join(out)

    def line(self, x1, y1, x2, y2, colour="ink", *, width=1.4, arrow=True, dash="") -> str:
        stroke, _ = PALETTE[colour]
        extra = f' stroke-dasharray="{dash}"' if dash else ""
        out = [
            f'<path d="{self._edge(x1, y1, x2, y2, bow=2.2)}" fill="none" '
            f'stroke="{stroke}" stroke-width="{width}" stroke-linecap="round"{extra}/>'
            for _pass in range(2)
        ]
        if arrow:
            out.append(self._arrowhead(x1, y1, x2, y2, stroke, width))
        return "".join(out)

    def curve(self, points, colour="ink", *, width=1.4, arrow=True, dash="") -> str:
        """A drawn-through polyline: consecutive bowed edges."""
        stroke, _ = PALETTE[colour]
        extra = f' stroke-dasharray="{dash}"' if dash else ""
        out = []
        for _pass in range(2):
            d = " ".join(
                self._edge(*points[i], *points[i + 1], bow=2.4)
                for i in range(len(points) - 1)
            )
            out.append(
                f'<path d="{d}" fill="none" stroke="{stroke}" '
                f'stroke-width="{width}" stroke-linecap="round"{extra}/>'
            )
        if arrow:
            out.append(self._arrowhead(*points[-2], *points[-1], stroke, width))
        return "".join(out)

    def _arrowhead(self, x1, y1, x2, y2, stroke, width) -> str:
        """Two strokes off the tip — drawn, not a filled triangle marker."""
        a = math.atan2(y2 - y1, x2 - x1)
        size = 7.5
        out = []
        for turn in (a + 2.55, a - 2.55):
            ex = x2 + size * math.cos(turn) + self.j(0.7)
            ey = y2 + size * math.sin(turn) + self.j(0.7)
            out.append(
                f'<path d="M{x2:.1f},{y2:.1f} L{ex:.1f},{ey:.1f}" fill="none" '
                f'stroke="{stroke}" stroke-width="{width}" stroke-linecap="round"/>'
            )
        return "".join(out)

    def cross(self, cx, cy, r=11, colour="red", *, width=2.4) -> str:
        stroke, _ = PALETTE[colour]
        return "".join(
            f'<path d="{self._edge(cx - r, cy - r, cx + r, cy + r, 1.4)}" fill="none" '
            f'stroke="{stroke}" stroke-width="{width}" stroke-linecap="round"/>'
            f'<path d="{self._edge(cx + r, cy - r, cx - r, cy + r, 1.4)}" fill="none" '
            f'stroke="{stroke}" stroke-width="{width}" stroke-linecap="round"/>'
            for _ in range(1)
        )


def text(x, y, s, *, size=8.4, colour="#454c53", weight="normal", anchor="start",
         family=HAND) -> str:
    esc = s.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
    w = f' font-weight="{weight}"' if weight != "normal" else ""
    a = f' text-anchor="{anchor}"' if anchor != "start" else ""
    return (
        f'<text x="{x}" y="{y}" font-size="{size}" fill="{colour}" '
        f'font-family="{family}"{w}{a}>{esc}</text>'
    )


def mono(x, y, s, **kw) -> str:
    kw.setdefault("family", "DejaVu Sans Mono, monospace")
    return text(x, y, s, **kw)


def figure(width, height, body) -> str:
    return (
        f'<svg viewBox="0 0 {width} {height}" width="{width}" height="{height}">'
        f"{body}</svg>"
    )

File: vc-agent/pipeline/web/rough.js
/* Hand-drawn SVG primitives — the browser half of docs/diagrams/rough.py.
 *
 * The same three rules make a shape read as drawn rather than plotted, and the
 * third is the one people miss:
 *
 *   1. Corners miss      — every vertex moves a pixel or two
 *   2. Edges bow         — each edge is a quadratic curve with a small bulge
 *   3. Every stroke twice — with different jitter. A pen goes back over a line,
 *                           and this contributes more than the other two together
 *
 * Why a port rather than server-rendered SVG: this panel *highlights* — the
 * active node thickens as a turn moves through it — so the shapes have to be
 * redrawn in place. Shipping a static picture per state would mean shipping one
 * picture per state.
 *
 * Jitter is seeded per shape, so a redraw of the same diagram is identical. A
 * panel that reshuffles its own wobble on every repaint looks like it is doing
 * something, which is exactly the wrong signal from a monitor.
 */
(function (global) {
  'use strict';

  var NS = 'http://www.w3.org/2000/svg';

  /* Mulberry32 — small, fast, and deterministic from a single integer. */
  function seeded(seed) {
    var a = seed >>> 0;
    return function () {
      a = (a + 0x6D2B79F5) >>> 0;
      var t = Math.imul(a ^ (a >>> 15), 1 | a);
      t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
      return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
    };
  }

  function Pen(seed, roughness) {
    this.rand = seeded(seed || 1);
    this.roughness = roughness == null ? 1 : roughness;
  }

  Pen.prototype.j = function (amount) {
    var a = amount == null ? 1.6 : amount;
    return (this.rand() * 2 - 1) * a * this.roughness;
  };

  /* One bowed edge. The bulge is perpendicular to the line, never along it —
     along it just makes the line look mis-measured rather than hand-drawn. */
  Pen.prototype.edge = function (x1, y1, x2, y2, bow) {
    var dx = x2 - x1, dy = y2 - y1;
    var len = Math.hypot(dx, dy) || 1;
    var nx = -dy / len, ny = dx / len;
    var swell = this.j(bow == null ? 1.8 : bow) * Math.min(1, len / 90);
    var mx = (x1 + x2) / 2 + nx * swell;
    var my = (y1 + y2) / 2 + ny * swell;
    return 'M' + (x1 + this.j()).toFixed(1) + ',' + (y1 + this.j()).toFixed(1) +
           ' Q' + mx.toFixed(1) + ',' + my.toFixed(1) +
           ' ' + (x2 + this.j()).toFixed(1) + ',' + (y2 + this.j()).toFixed(1);
  };

  function path(d, stroke, width, extra) {
    var node = document.createElementNS(NS, 'path');
    node.setAttribute('d', d);
    node.setAttribute('fill', 'none');
    node.setAttribute('stroke', stroke);
    node.setAttribute('stroke-width', String(width));
    node.setAttribute('stroke-linecap', 'round');
    if (extra) { node.setAttribute('stroke-dasharray', extra); }
    return node;
  }

  Pen.prototype.rect = function (x, y, w, h, opts) {
    var o = opts || {};
    var g = document.createElementNS(NS, 'g');
    if (o.fill) {
      var wash = document.createElementNS(NS, 'path');
      wash.setAttribute('d', 'M' + (x + 1) + ',' + (y + 1) + ' L' + (x + w - 1) + ',' + (y + 1) +
                             ' L' + (x + w - 1) + ',' + (y + h - 1) + ' L' + (x + 1) + ',' + (y + h - 1) + ' Z');
      wash.setAttribute('fill', o.fill);
      wash.setAttribute('stroke', 'none');
      g.appendChild(wash);
    }
    var corners = [[x, y], [x + w, y], [x + w, y + h], [x, y + h]];
    for (var pass = 0; pass < 2; pass++) {
      var d = '';
      for (var i = 0; i < 4; i++) {
        var a = corners[i], b = corners[(i + 1) % 4];
        d += this.edge(a[0], a[1], b[0], b[1], 1.8) + ' ';
      }
      g.appendChild(path(d, o.stroke || '#1e1e1e', o.width || 1.6, o.dash));
    }
    return g;
  };

  Pen.prototype.line = function (x1, y1, x2, y2, opts) {
    var o = opts || {};
    var g = document.createElementNS(NS, 'g');
    for (var pass = 0; pass < 2; pass++) {
      g.appendChild(path(this.edge(x1, y1, x2, y2, 2.2), o.stroke || '#1e1e1e', o.width || 1.4, o.dash));
    }
    if (o.arrow !== false) { g.appendChild(this.arrowhead(x1, y1, x2, y2, o)); }
    return g;
  };

  Pen.prototype.curve = function (points, opts) {
    var o = opts || {};
    var g = document.createElementNS(NS, 'g');
    for (var pass = 0; pass < 2; pass++) {
      var d = '';
      for (var i = 0; i < points.length - 1; i++) {
        d += this.edge(points[i][0], points[i][1], points[i + 1][0], points[i + 1][1], 2.4) + ' ';
      }
      g.appendChild(path(d, o.stroke || '#1e1e1e', o.width || 1.4, o.dash));
    }
    if (o.arrow !== false) {
      var n = points.length;
      g.appendChild(this.arrowhead(points[n - 2][0], points[n - 2][1], points[n - 1][0], points[n - 1][1], o));
    }
    return g;
  };

  /* Two strokes off the tip, not a filled marker — a drawn arrow has an open head. */
  Pen.prototype.arrowhead = function (x1, y1, x2, y2, opts) {
    var o = opts || {};
    var a = Math.atan2(y2 - y1, x2 - x1);
    var g = document.createElementNS(NS, 'g');
    var turns = [a + 2.55, a - 2.55];
    for (var i = 0; i < turns.length; i++) {
      var ex = x2 + 7.5 * Math.cos(turns[i]) + this.j(0.7);
      var ey = y2 + 7.5 * Math.sin(turns[i]) + this.j(0.7);
      g.appendChild(path('M' + x2.toFixed(1) + ',' + y2.toFixed(1) + ' L' + ex.toFixed(1) + ',' + ey.toFixed(1),
                         o.stroke || '#1e1e1e', o.width || 1.4));
    }
    return g;
  };

  function text(x, y, s, opts) {
    var o = opts || {};
    var node = document.createElementNS(NS, 'text');
    node.setAttribute('x', String(x));
    node.setAttribute('y', String(y));
    node.setAttribute('font-size', String(o.size || 8.4));
    node.setAttribute('fill', o.colour || '#454c53');
    node.setAttribute('font-family', o.mono
      ? 'ui-monospace, SFMono-Regular, Menlo, monospace'
      : '"Comic Sans MS", "Comic Neue", system-ui, sans-serif');
    if (o.weight) { node.setAttribute('font-weight', o.weight); }
    if (o.anchor) { node.setAttribute('text-anchor', o.anchor); }
    node.textContent = s;
    return node;
  }

  function svg(width, height) {
    var node = document.createElementNS(NS, 'svg');
    node.setAttribute('viewBox', '0 0 ' + width + ' ' + height);
    node.setAttribute('width', String(width));
    node.setAttribute('height', String(height));
    return node;
  }

  global.Rough = { Pen: Pen, text: text, svg: svg, NS: NS };
})(window);
